lib/auth/github.go

```diff
@@ -84,6 +91,124 @@ func (a *Server) upsertGithubConnector(ctx context.Context, connector types.Gith
 	return nil
 }
 
+// HTTPRequester allows a net/http.Client to be mocked for tests.
+type HTTPRequester interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// checkGithubOrgSSOSupport returns an error if any of the Github
+// organizations specified in this connector use external SSO.
+// If userTeams is not nil, only organizations that are both specified
+// in conn and in userTeams will be checked. If client is nil a
+// net/http.Client will be used.
+func checkGithubOrgSSOSupport(ctx context.Context, conn types.GithubConnector, userTeams []teamResponse, orgCache *utils.FnCache, client HTTPRequester) error {
+	version := modules.GetModules().BuildType()
+	if version == modules.BuildEnterprise {
+		return nil
+	}
+
+	orgs := make(map[string]struct{})
+	addOrg := func(org string) {
+		if len(userTeams) != 0 {
+			// Only check organizations that the user is a member of and
+			// that are specified in this auth connector
+			for _, team := range userTeams {
+				if org == team.Org.Login {
+					orgs[org] = struct{}{}
+				}
+			}
+		} else {
+			orgs[org] = struct{}{}
+		}
+	}
+
+	// Check each organization only once
+	for _, mapping := range conn.GetTeamsToLogins() {
+		addOrg(mapping.Organization)
+	}
+	for _, mapping := range conn.GetTeamsToRoles() {
+		addOrg(mapping.Organization)
+	}
+
+	if client == nil {
+		var err error
+		client, err = defaults.HTTPClient()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	for org := range orgs {
+		orgResult, err := orgCache.Get(ctx, org, func(ctx context.Context) (interface{}, error) {
+			return orgUsesExternalSSO(ctx, org, client)
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		usesSSO, ok := orgResult.(bool)
+		if !ok {
+			return trace.BadParameter("Expected bool from cache, got %T", orgResult)
+		}
+		if usesSSO {
+			return trace.AccessDenied(
+				"GitHub organization %s uses external SSO, please purchase a Teleport Enterprise license if you want to authenticate with this organization",
+				org,
+			)
+		}
+	}
+
+	return nil
+}
+
+// orgUsesExternalSSO returns true if the Github organization org
+// uses external SSO.
+func orgUsesExternalSSO(ctx context.Context, org string, client HTTPRequester) (bool, error) {
+	// A Github organization will have a "sso" page reachable if it
+	// supports external SSO. There doesn't seem to be any way to get this
+	// information from the Github REST API without being an owner of the
+	// Github organization, so check if this exists instead.
+	ssoURL := fmt.Sprintf("%s/%s/sso", githubOrgsURL, url.PathEscape(org))
+
+	const retries = 3
+	var resp *http.Response
+	for i := 0; i < retries; i++ {
+		ctx, cancel := context.WithTimeout(ctx, defaults.HTTPRequestTimeout)
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, ssoURL, nil)
+		cancel()
```

Comment:
Why are you canceling the context immediately? I would expect this to fail every time because you haven't even made the request yet.

Reply:
:man_facepalming: that was a brain fart, moved into a function like @jakule originally suggested